test/tstTreeHelpers.cpp

```diff
@@ -0,0 +1,199 @@
+}
+
+} // namespace Test
+
+#define ARBORX_TEST_REDUCE_LABELS(exec_space, parents, labels, ref)            \
+  BOOST_TEST(Test::reduceLabels(exec_space, parents, labels) == ref,           \
+             boost::test_tools::per_element())
+
+BOOST_AUTO_TEST_SUITE(TreeHelpers)
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(reduce_labels, DeviceType, ARBORX_DEVICE_TYPES)
+{
+  /*
+      [0]------------*--------------
+                    / \
+       ------*----[3] [4]*----------
+            / \         / \
+       --*[1] [2]*--   |  [5]----*--
+        / \     / \    |        / \
+       |   |   |   |   |   --*[6]  |
+       |   |   |   |   |    / \    |
+       0   1   2   3   4   5   6   7
+  */
+  auto const parents =
+      std::vector<int>{-1, 3, 3, 0, 0, 4, 5, 1, 1, 2, 2, 4, 6, 6, 5};
+  //                    0  1  2  3  4  5  6  7  8  9 10 11 12 13 14
+  //                   [0][1][2][3][4][5][6] 0  1  2  3  4  5  6  7
+
+  using ExecutionSpace = typename DeviceType::execution_space;
+  ExecutionSpace exec_space;
+
+  ARBORX_TEST_REDUCE_LABELS(
+      exec_space, parents,
+      (std::vector<int>{0, 1, 2, 3, 4, 5, 6, 0, 0, 0, 0, 0, 0, 0, 0}),
+      (std::vector<int>{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}));
+
+  ARBORX_TEST_REDUCE_LABELS(
+      exec_space, parents,
+      (std::vector<int>{0, 1, 2, 3, 4, 5, 6, 7, 7, 7, 7, 7, 7, 7, 7}),
+      (std::vector<int>{7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7}));
+
+  ARBORX_TEST_REDUCE_LABELS(
+      exec_space, parents,
+      (std::vector<int>{0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6, 7}),
+      (std::vector<int>{-1, -1, -1, -1, -1, -1, -1, 0, 1, 2, 3, 4, 5, 6, 7}));
+
+  ARBORX_TEST_REDUCE_LABELS(
+      exec_space, parents,
+      (std::vector<int>{0, 1, 2, 3, 4, 5, 6, 0, 0, 0, 3, 4, 4, 4, 7}),
+      (std::vector<int>{-1, 0, -1, -1, -1, -1, 4, 0, 0, 0, 3, 4, 4, 4, 7}));
+}
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(find_parents, DeviceType, ARBORX_DEVICE_TYPES)
+{
+  using ExecutionSpace = typename DeviceType::execution_space;
+  using MemorySpace = typename DeviceType::memory_space;
+
+  ExecutionSpace exec_space;
+
+  int const n = 103;
+
+  auto bvh = Test::buildRandomBVH(exec_space, n);
+
+  Kokkos::View<int *, MemorySpace> bvh_parents("Test::bvh_parents", 2 * n - 1);
+  Kokkos::deep_copy(exec_space, bvh_parents, -1);
+
+  ArborX::Experimental::findBVHParents(exec_space, bvh, bvh_parents);
+
+  int num_failures = 0;
+  Kokkos::parallel_reduce(
+      "Test::check_bvh_parents",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, 2 * n - 1),
+      KOKKOS_LAMBDA(int i, int &update) {
+        using ArborX::Details::HappyTreeFriends;
+
+        int const parent = bvh_parents(i);
+
+        if (i == 0) // root
+        {
+          if (parent != -1)
+            ++update;
+          return;
+        }
+
+        if (parent < 0 || parent >= n)
+        {
+          ++update;
+          return;
+        }
+
+        if (HappyTreeFriends::getLeftChild(bvh, parent) != i &&
```

Comment:
```suggestion
        if (HappyTreeFriends::getLeftChild(bvh, parent) != i ||
```
ouch

Reply:
No. To update the error count it must not much either of the children.

Reply:
Let's not use English to litigate because we are both bad at it.
Consider the tree below
```
[0]----*----
      / \
 --*[1]  |
  / \    |
 |   |   |
 |   |   |
 0   1   2
```
The right answer is `{-1, 0, 1, 1, 0}` but `{-1, 0, 1, 1, 1}` for instance would pass the test as it is currently written.

Reply:
You want to assert that 
`(getLeftChild(bvh, parent) == i) && (getRIghtChild(bvh, parent) == i)`
so the condition should read `!((getLeftChild(bvh, parent) == i) && (getRIghtChild(bvh, parent) == i))`
which is equivalent to `(getLeftChild(bvh, parent) != i) || (getRIghtChild(bvh, parent) != i)`

Reply:
Andrey pointed out that the condition to assert is `(getLeftChild(bvh, parent) == i) || (getRIghtChild(bvh, parent) == i)`